File: code/tests/cases/test_error.m
/*
 * -----------------------------------------------------------------------------
 * Project: Fossil Logic
 *
 * This file is part of the Fossil Logic project, which aims to develop high-
 * performance, cross-platform applications and libraries. The code contained
 * herein is subject to the terms and conditions defined in the project license.
 *
 * Author: Michael Gene Brockus (Dreamer)
 *
 * Copyright (C) 2024 Fossil Logic. All rights reserved.
 * -----------------------------------------------------------------------------
 */
#include <fossil/pizza/framework.h>

#include "fossil/io/framework.h"

// * * * * * * * * * * * * * * * * * * * * * * * *
// * Fossil Logic Test Utilites
// * * * * * * * * * * * * * * * * * * * * * * * *
// Setup steps for things like test fixtures and
// mock objects are set here.
// * * * * * * * * * * * * * * * * * * * * * * * *

// Define the test suite and add test cases
FOSSIL_SUITE(objc_error_suite);

// Setup function for the test suite
FOSSIL_SETUP(objc_error_suite) {
    // Setup code here
}

// Teardown function for the test suite
FOSSIL_TEARDOWN(objc_error_suite) {
    // Teardown code here
}

// * * * * * * * * * * * * * * * * * * * * * * * *
// * Fossil Logic Test Cases
// * * * * * * * * * * * * * * * * * * * * * * * *
// The test cases below are provided as samples, inspired
// by the Meson build system's approach of using test cases
// as samples for library usage.
// * * * * * * * * * * * * * * * * * * * * * * * *

FOSSIL_TEST(objc_test_io_what_no_error) {
    const char *result = fossil_io_what(FOSSIL_ERROR_OK);
    ASSUME_ITS_EQUAL_CSTR("No error, operation successful.", result);
}

FOSSIL_TEST(objc_test_io_what_CNULL_pointer) {
    const char *result = fossil_io_what(FOSSIL_ERROR_CNULL_POINTER);
    ASSUME_ITS_EQUAL_CSTR("Null pointer encountered.", result);
}

FOSSIL_TEST(objc_test_io_what_invalid_argument) {
    const char *result = fossil_io_what(FOSSIL_ERROR_INVALID_ARGUMENT);
    ASSUME_ITS_EQUAL_CSTR("Invalid argument provided.", result);
}

FOSSIL_TEST(objc_test_io_what_type_mismatch) {
    const char *result = fossil_io_what(FOSSIL_ERROR_TYPE_MISMATCH);
    ASSUME_ITS_EQUAL_CSTR("Type mismatch encountered.", result);
}

FOSSIL_TEST(objc_test_io_what_invalid_operation) {
    const char *result = fossil_io_what(FOSSIL_ERROR_INVALID_OPERATION);
    ASSUME_ITS_EQUAL_CSTR("Invalid operation.", result);
}

FOSSIL_TEST(objc_test_io_what_unknown) {
    const char *result = fossil_io_what(FOSSIL_ERROR_UNKNOWN);
    ASSUME_ITS_EQUAL_CSTR("Unknown error.", result);
}

FOSSIL_TEST(objc_test_io_what_custom) {
    const char *result = fossil_io_what(FOSSIL_ERROR_CUSTOM);
    ASSUME_ITS_EQUAL_CSTR("Custom error occurred.", result);
}

FOSSIL_TEST(objc_test_io_what_internal) {
    const char *result = fossil_io_what(FOSSIL_ERROR_INTERNAL);
    ASSUME_ITS_EQUAL_CSTR("Internal error.", result);
}

FOSSIL_TEST(objc_test_io_what_unknown_error_code) {
    const char *result = fossil_io_what(FOSSIL_ERROR_UNKNOWN_ERROR_CODE);
    ASSUME_ITS_EQUAL_CSTR("Unknown error code.", result);
}

FOSSIL_TEST(objc_test_io_what_overflow_int) {
    const char *result = fossil_io_what(FOSSIL_ERROR_OVERFLOW_INT);
    ASSUME_ITS_EQUAL_CSTR("Integer overflow.", result);
}

FOSSIL_TEST(objc_test_io_what_underflow_int) {
    const char *result = fossil_io_what(FOSSIL_ERROR_UNDERFLOW_INT);
    ASSUME_ITS_EQUAL_CSTR("Integer underflow.", result);
}

FOSSIL_TEST(objc_test_io_what_overflow_float) {
    const char *result = fossil_io_what(FOSSIL_ERROR_OVERFLOW_FLOAT);
    ASSUME_ITS_EQUAL_CSTR("Float overflow.", result);
}

FOSSIL_TEST(objc_test_io_what_underflow_float) {
    const char *result = fossil_io_what(FOSSIL_ERROR_UNDERFLOW_FLOAT);
    ASSUME_ITS_EQUAL_CSTR("Float underflow.", result);
}

FOSSIL_TEST(objc_test_io_what_division_by_zero) {
    const char *result = fossil_io_what(FOSSIL_ERROR_DIVISION_BY_ZERO);
    ASSUME_ITS_EQUAL_CSTR("Division by zero.", result);
}

FOSSIL_TEST(objc_test_io_what_invalid_cast) {
    const char *result = fossil_io_what(FOSSIL_ERROR_INVALID_CAST);
    ASSUME_ITS_EQUAL_CSTR("Invalid type cast.", result);
}

FOSSIL_TEST(objc_test_io_what_out_of_memory) {
    const char *result = fossil_io_what(FOSSIL_ERROR_OUT_OF_MEMORY);
    ASSUME_ITS_EQUAL_CSTR("Out of memory.", result);
}

FOSSIL_TEST(objc_test_io_what_memory_corruption) {
    const char *result = fossil_io_what(FOSSIL_ERROR_MEMORY_CORRUPTION);
    ASSUME_ITS_EQUAL_CSTR("Memory corruption detected.", result);
}

FOSSIL_TEST(objc_test_io_what_buffer_overflow) {
    const char *result = fossil_io_what(FOSSIL_ERROR_BUFFER_OVERFLOW);
    ASSUME_ITS_EQUAL_CSTR("Buffer overflow.", result);
}

FOSSIL_TEST(objc_test_io_what_buffer_underflow) {
    const char *result = fossil_io_what(FOSSIL_ERROR_BUFFER_UNDERFLOW);
    ASSUME_ITS_EQUAL_CSTR("Buffer underflow.", result);
}

FOSSIL_TEST(objc_test_io_what_file_not_found) {
    const char *result = fossil_io_what(FOSSIL_ERROR_FILE_NOT_FOUND);
    ASSUME_ITS_EQUAL_CSTR("File not found.", result);
}

FOSSIL_TEST(objc_test_io_what_permission_denied) {
    const char *result = fossil_io_what(FOSSIL_ERROR_PERMISSION_DENIED);
    ASSUME_ITS_EQUAL_CSTR("Permission denied.", result);
}

FOSSIL_TEST(objc_test_io_what_network_failure) {
    const char *result = fossil_io_what(FOSSIL_ERROR_NETWORK_FAILURE);
    ASSUME_ITS_EQUAL_CSTR("Network failure.", result);
}

FOSSIL_TEST(objc_test_io_what_timeout) {
    const char *result = fossil_io_what(FOSSIL_ERROR_TIMEOUT);
    ASSUME_ITS_EQUAL_CSTR("Network timeout.", result);
}

FOSSIL_TEST(objc_test_io_what_unknown_host) {
    const char *result = fossil_io_what(FOSSIL_ERROR_UNKNOWN_HOST);
    ASSUME_ITS_EQUAL_CSTR("Unknown host.", result);
}

FOSSIL_TEST(objc_test_io_what_connection_refused) {
    const char *result = fossil_io_what(FOSSIL_ERROR_CONNECTION_REFUSED);
    ASSUME_ITS_EQUAL_CSTR("Connection refused.", result);
}

FOSSIL_TEST(objc_test_io_what_sql_injection) {
    const char *result = fossil_io_what(FOSSIL_ERROR_SQL_INJECTION);
    ASSUME_ITS_EQUAL_CSTR("SQL injection attempt detected.", result);
}

FOSSIL_TEST(objc_test_io_what_xss_attack) {
    const char *result = fossil_io_what(FOSSIL_ERROR_XSS_ATTACK);
    ASSUME_ITS_EQUAL_CSTR("Cross-site scripting attack detected.", result);
}

FOSSIL_TEST(objc_test_io_what_csrf_attack) {
    const char *result = fossil_io_what(FOSSIL_ERROR_CSRF_ATTACK);
    ASSUME_ITS_EQUAL_CSTR("Cross-site request forgery attack detected.", result);
}

FOSSIL_TEST(objc_test_io_what_user_abort) {
    const char *result = fossil_io_what(FOSSIL_ERROR_USER_ABORT);
    ASSUME_ITS_EQUAL_CSTR("User aborted operation.", result);
}

FOSSIL_TEST(objc_test_io_what_database_connection_failed) {
    const char *result = fossil_io_what(FOSSIL_ERROR_DATABASE_CONNECTION_FAILED);
    ASSUME_ITS_EQUAL_CSTR("Database connection failed.", result);
}

FOSSIL_TEST(objc_test_io_what_serialization_failed) {
    const char *result = fossil_io_what(FOSSIL_ERROR_SERIALIZATION_FAILED);
    ASSUME_ITS_EQUAL_CSTR("Serialization failed.", result);
}

// * * * * * * * * * * * * * * * * * * * * * * * *
// * Fossil Logic Test Pool
// * * * * * * * * * * * * * * * * * * * * * * * *

FOSSIL_TEST_GROUP(objc_error_tests) {
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_no_error);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_CNULL_pointer);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_invalid_argument);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_type_mismatch);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_invalid_operation);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_unknown);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_custom);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_internal);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_unknown_error_code);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_overflow_int);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_underflow_int);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_overflow_float);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_underflow_float);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_division_by_zero);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_invalid_cast);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_out_of_memory);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_memory_corruption);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_buffer_overflow);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_buffer_underflow);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_file_not_found);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_permission_denied);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_network_failure);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_timeout);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_unknown_host);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_connection_refused);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_sql_injection);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_xss_attack);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_csrf_attack);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_user_abort);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_database_connection_failed);
    FOSSIL_TEST_ADD(objc_error_suite, objc_test_io_what_serialization_failed);

    FOSSIL_TEST_REGISTER(objc_error_suite);
}

File: code/tests/cases/test_error.cpp
/**
 * -----------------------------------------------------------------------------
 * Project: Fossil Logic
 *
 * This file is part of the Fossil Logic project, which aims to develop
 * high-performance, cross-platform applications and libraries. The code
 * contained herein is licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain
 * a copy of the License at:
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * Author: Michael Gene Brockus (Dreamer)
 * Date: 04/05/2014
 *
 * Copyright (C) 2014-2025 Fossil Logic. All rights reserved.
 * -----------------------------------------------------------------------------
 */
#include <fossil/pizza/framework.h>

#include "fossil/io/framework.h"

// * * * * * * * * * * * * * * * * * * * * * * * *
// * Fossil Logic Test Utilites
// * * * * * * * * * * * * * * * * * * * * * * * *
// Setup steps for things like test fixtures and
// mock objects are set here.
// * * * * * * * * * * * * * * * * * * * * * * * *

// Define the test suite and add test cases
FOSSIL_SUITE(cpp_error_suite);

// Setup function for the test suite
FOSSIL_SETUP(cpp_error_suite) {
    // Setup code here
}

// Teardown function for the test suite
FOSSIL_TEARDOWN(cpp_error_suite) {
    // Teardown code here
}

// * * * * * * * * * * * * * * * * * * * * * * * *
// * Fossil Logic Test Cases
// * * * * * * * * * * * * * * * * * * * * * * * *
// The test cases below are provided as samples, inspired
// by the Meson build system's approach of using test cases
// as samples for library usage.
// * * * * * * * * * * * * * * * * * * * * * * * *

// SYSTEM / META error code tests
FOSSIL_TEST(cpp_test_error_code_system_meta) {
    const char *codes[] = {
        "system.ok", "system.unknown", "system.internal", "system.fatal", "system.panic",
        "system.abort", "system.assertion", "system.invariant", "system.contract",
        "system.recoverable", "system.unrecoverable", "system.transient", "system.permanent",
        "system.unsupported", "system.unimplemented", "system.deprecated", "system.experimental",
        "system.misconfigured", "system.corrupt", "system.bootstrap", "system.shutdown",
        "system.restart", "system.upgrade", "system.downgrade", "system.permission",
        "system.capability"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// IO error code tests
FOSSIL_TEST(cpp_test_error_code_io) {
    const char *codes[] = {
        "io.read", "io.write", "io.seek", "io.flush", "io.sync", "io.fsync", "io.truncate",
        "io.append", "io.scatter", "io.gather", "io.closed", "io.eof", "io.partial",
        "io.short", "io.blocked", "io.nonblocking", "io.timeout", "io.interrupt", "io.retry",
        "io.corrupt", "io.checksum", "io.buffer", "io.alignment", "io.direct", "io.stream",
        "io.pipe"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// MEMORY error code tests
FOSSIL_TEST(cpp_test_error_code_memory) {
    const char *codes[] = {
        "memory.alloc", "memory.realloc", "memory.free", "memory.map", "memory.unmap",
        "memory.remap", "memory.lock", "memory.unlock", "memory.protect", "memory.unprotect",
        "memory.leak", "memory.overrun", "memory.underrun", "memory.use_after_free",
        "memory.double_free", "memory.fragmented", "memory.exhausted", "memory.alignment",
        "memory.page_fault", "memory.segmentation", "memory.guard", "memory.poison",
        "memory.swap", "memory.numa"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// CPU / EXECUTION error code tests
FOSSIL_TEST(cpp_test_error_code_cpu) {
    const char *codes[] = {
        "cpu.illegal_instruction", "cpu.privilege_violation", "cpu.div_zero", "cpu.overflow",
        "cpu.underflow", "cpu.fpu", "cpu.simd", "cpu.cache", "cpu.pipeline", "cpu.affinity",
        "cpu.throttle"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// MATH / NUMERIC error code tests
FOSSIL_TEST(cpp_test_error_code_math) {
    const char *codes[] = {
        "math.overflow", "math.underflow", "math.div_zero", "math.nan", "math.infinity",
        "math.domain", "math.range", "math.precision", "math.rounding", "math.convergence",
        "math.divergence", "math.iteration", "math.singularity", "math.condition",
        "math.approximation"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// FILESYSTEM / STORAGE error code tests
FOSSIL_TEST(cpp_test_error_code_fs) {
    const char *codes[] = {
        "fs.not_found", "fs.exists", "fs.permission", "fs.read_only", "fs.locked",
        "fs.busy", "fs.mount", "fs.unmount", "fs.remount", "fs.quota", "fs.corrupt",
        "fs.journal", "fs.snapshot", "fs.backup", "fs.restore", "fs.path", "fs.symlink",
        "fs.hardlink", "fs.inode", "fs.filesystem"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// NETWORK error code tests
FOSSIL_TEST(cpp_test_error_code_network) {
    const char *codes[] = {
        "network.unreachable", "network.timeout", "network.reset", "network.refused",
        "network.aborted", "network.dns", "network.routing", "network.latency",
        "network.bandwidth", "network.congestion", "network.proxy", "network.firewall",
        "network.nat", "network.session", "network.stream", "network.packet"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// SECURITY / CRYPTO error code tests
FOSSIL_TEST(cpp_test_error_code_security) {
    const char *codes[] = {
        "security.violation", "security.auth_failed", "security.authz_failed",
        "security.identity", "security.credential", "security.token", "security.session",
        "security.encryption", "security.decryption", "security.certificate", "security.key",
        "security.keystore", "security.revocation", "security.sandbox", "security.trust",
        "security.integrity", "security.tamper", "security.replay"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// DATABASE / STORAGE ENGINE error code tests
FOSSIL_TEST(cpp_test_error_code_database) {
    const char *codes[] = {
        "database.connect", "database.disconnect", "database.query", "database.prepare",
        "database.execute", "database.transaction", "database.commit", "database.rollback",
        "database.deadlock", "database.lock", "database.constraint", "database.schema",
        "database.migration", "database.index", "database.cursor", "database.replication",
        "database.consistency", "database.timeout"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// AI / ML error code tests
FOSSIL_TEST(cpp_test_error_code_ai) {
    const char *codes[] = {
        "ai.model", "ai.version", "ai.load", "ai.unload", "ai.inference", "ai.training",
        "ai.finetune", "ai.dataset", "ai.validation", "ai.bias", "ai.drift",
        "ai.hallucination", "ai.alignment", "ai.confidence", "ai.explainability",
        "ai.prompt", "ai.token_limit", "ai.context_overflow"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// UI / UX error code tests
FOSSIL_TEST(cpp_test_error_code_ui) {
    const char *codes[] = {
        "ui.render", "ui.layout", "ui.paint", "ui.refresh", "ui.input", "ui.focus",
        "ui.gesture", "ui.accessibility", "ui.localization", "ui.theme", "ui.font",
        "ui.image"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// LEGAL / POLICY error code tests
FOSSIL_TEST(cpp_test_error_code_policy) {
    const char *codes[] = {
        "policy.violation", "policy.denied", "policy.expired", "policy.restricted",
        "license.invalid", "license.expired", "license.restricted", "privacy.violation",
        "privacy.redaction", "compliance.failure", "audit.failure"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// Fallback / meta error code tests
FOSSIL_TEST(cpp_test_error_code_meta) {
    const char *codes[] = {
        "meta.unreachable", "meta.assumption", "meta.placeholder", "meta.future"
    };
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        int id = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(id >= 0);
        const char *msg = fossil::io::Error::what(codes[i]);
        ASSUME_ITS_TRUE(msg && strcmp(msg, "unknown") != 0);
    }
}

// Test that all codes are unique and stable
FOSSIL_TEST(cpp_test_error_code_uniqueness_and_stability) {
    const char *codes[] = {
        "system.ok", "io.read", "memory.alloc", "cpu.div_zero", "math.nan", "parse.syntax",
        "type.mismatch", "format.invalid", "data.corrupt", "fs.not_found", "process.spawn",
        "thread.create", "resource.exhausted", "time.timeout", "config.missing",
        "api.invalid_call", "protocol.invalid", "network.unreachable", "security.violation",
        "database.connect", "ai.model", "ui.render", "log.write", "build.compile",
        "user.input", "policy.violation", "meta.unreachable"
    };
    int ids[sizeof(codes)/sizeof(*codes)];
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        ids[i] = fossil::io::Error::code(codes[i]);
        ASSUME_ITS_TRUE(ids[i] >= 0);
    }
    for (size_t i = 0; i < sizeof(codes)/sizeof(*codes); ++i) {
        for (size_t j = i + 1; j < sizeof(codes)/sizeof(*codes); ++j) {
            ASSUME_ITS_TRUE(ids[i] != ids[j]);
        }
    }
}

// * * * * * * * * * * * * * * * * * * * * * * * *
// * Fossil Logic Test Pool
// * * * * * * * * * * * * * * * * * * * * * * * *

FOSSIL_TEST_GROUP(cpp_error_tests) {
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_system_meta);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_io);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_memory);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_cpu);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_math);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_fs);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_network);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_security);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_database);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_ai);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_ui);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_policy);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_meta);
    FOSSIL_TEST_ADD(cpp_error_suite, cpp_test_error_code_uniqueness_and_stability);

    FOSSIL_TEST_REGISTER(cpp_error_suite);
}
